### ISSUE TITLE
Optimize bucket_ref() to not make _bucket:get() when _bucket does not change

### ISSUE THREAD

**Issue by Gerold103** [tarantool/vshard#285]
`bucket_ref()` is used by `storage_call()` which is used by all the routers most of the time. It has an optimization that if there is already > 0 refs of the requested type, it works as an increment. But if the current ref count is 0, it makes `_bucket:get()` to check bucket state. That might be expensive - a tuple is returned to Lua as cdata, which increases GC pressure. Also it might be relatively slow if bucket count is millions.

Instead, `bucket_ref()` could proceed to the ref increment even if its current ref count is 0. For that the `_bucket` space's trigger could remove the ref of the changed bucket if its ref count is 0. Then, if `bucket_ref()` already sees a ref object, it can work with it without touching `_bucket`. If there is no a ref object, then work like now - get a tuple from `_bucket`, ensure it is suitable for the ref, create the ref object, etc.

Another option would be to save bucket generation into each `bucket_ref` object when it was changed last time. If on a next ref the generation is the same as the global one - nothing changed, can do the ref. The problem is that it will add +8 bytes for each ref (expensive on huge bucket count), and might not work well with hot reload.

**Comment by R-omk** [comment]
@Gerold103 
Why is this ticket closed?

For example, if the rebalancer is turned off, I would prefer that the ref system be completely disabled.

**Comment by Serpentian** [comment]
@R-omk 

The ref system doesn't add much overhead, since we no longer need to access the _bucket space during a ref. However, it provides significant protection and, in my opinion, should always be enabled—regardless of whether the rebalancer is active or not. It checks whether the bucket is present in storage and whether its state can be read or written.

The major problem with disabling refs, when rebalancer is disabled is that reconfiguration doesn't happen immediately on all replicasets and can fail in the middle. So, we may end up in the situation, where the rs with working rebalancer didn't get the configuration and rebalancer still works and the rest of the cluster already disabled the protection

I'm strongly against that approach, let's use the maximum protection, we can. Even if gives some overhead on the performace.

**Comment by Gerold103** [comment]
Yes, Nikita wrote it all, basically. The refs in their current shape cost for each request a one lookup in a Lua table + one `uint32_t` increment and decrement. Basically, nothing. The space-lookup happens once for each bucket. So after a short warmup the cost drops to neglectable values.

**Comment by R-omk** [comment]
@Gerold103 

ok, if you claim that reading from spaces is no longer used then everything is ok.

I drew attention to this because I have already encountered unexpected behavior when using transactions, and clearly emphasized that in different levels of isolation you can get unexpected results that can break the ref system



https://github.com/tarantool/vshard/issues/475#issuecomment-2250008141

**Comment by R-omk** [comment]
I looked at the current master and still see that

`bucket_refrw/ro` -> `if ref.rw == 0 then`  -> `bucket_check_state`  ->  `box.space._bucket:get({bucket_id})`


besides the fact that _bucket:get is a much more expensive operation compared to reading value from lua, it can also give an unpredictable result depending on whether the transaction is currently open and with what isolation level.

I think it need to cover with tests the scenarios when the ref is called inside a transaction.

**Comment by Serpentian** [comment]
> every time when ref couner in lua table eq zero

Just give us some time to cherry-pick commits, they'll be in master soon: https://github.com/tarantool/vshard/pull/553
